nixos/modules/services/web-servers/nginx/default.nix

```diff
@@ -4,21 +4,147 @@ with lib;
 
 let
   cfg = config.services.nginx;
-  nginx = cfg.package;
+  virtualHosts = mapAttrs (vhostName: vhostConfig:
+    vhostConfig // (optionalAttrs vhostConfig.enableACME {
+      sslCertificate = "/var/lib/acme/${vhostName}/fullchain.pem";
+      sslCertificateKey = "/var/lib/acme/${vhostName}/key.pem";
+    })
+  ) cfg.virtualHosts;
+
   configFile = pkgs.writeText "nginx.conf" ''
     user ${cfg.user} ${cfg.group};
+    error_log stderr;
     daemon off;
 
     ${cfg.config}
 
-    ${optionalString (cfg.httpConfig != "") ''
     http {
       include ${cfg.package}/conf/mime.types;
+      include ${cfg.package}/conf/fastcgi.conf;
+
+      # optimisation
+      sendfile on;
+      tcp_nopush on;
+      tcp_nodelay on;
+      keepalive_timeout 65;
+      types_hash_max_size 2048;
+
+      # use secure TLS defaults
+      ssl_protocols ${cfg.sslProtocols};
+      ssl_session_cache shared:SSL:42m;
+      ssl_session_timeout 23m;
+
+      ssl_ciphers ${cfg.sslCiphers};
+      ssl_ecdh_curve secp384r1;
```

Comment:
Though the P-256 curve must be supported, other curves can be and are supported as well. Unfortunately most TLS servers are not able to support multiple curves (nginx only with OpenSSL 1.0.2, see http://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_ecdh_curve).

All relevant browsers that support EC (including the newer IE versions) support secp384r1 & secp521r1 even before http2. As these curves were defined by NIST quite some time ago and have some proven weaknesses, there is some discussion about whether the NSA has artificially weakened them (see https://safecurves.cr.yp.to/). Not using the even weaker default here is just a healthy precaution in our opinion as better curves like ed25519 are still not available for ECDHE in TLS.

So yes, with that specific config without support for secp256r1 we're technically breaking the spec but it won't cause problems in practice.